### PR TITLE
Bugfix/ntfs_acl_permissions: Use GetAce() to enlist all ACE entries in dacl

### PR DESCRIPTION
**Pull request by ExceptionalHandler**
This is in response to the Bug :  #7113 

ntfs_acl_permissions table uses GetExplicitEntriesFromAcl() API which enumerates only explicit Aceess Conrol Entries (ACE) and not inherited ones. Most Windows files are protected by  inherited ACEs which will not get listed when running this table against those files. 

The PR intends to use GetACE() function to enum ACEs in a dacl which will get all ACEs regardless of whether they are explicit or not. The information returned by GetAce is slightly different than what GetExplicitEntriesFromAcl() might return
As a result, code to enumerate and populate column values needed to change though all efforts have beenn made to stick to existing column values. 

I have tested this on Windows 10. for both success and failure scenarios mentioned in #7113 
Sample output after fix:
```
PS E:\O\osquery\b> .\osquery\RelWithDebInfo\osqueryi.exe "select * from ntfs_acl_permissions where path = 'c:\users\administrator\desktop\tools\bginfo.exe'"
+-------------------------------------------------+-------+----------------+--------------------------------------------------------------------------------------------------------------+----------------+
| path                                            | type  | principal      | access                                                                                                       | inherited_from |
+-------------------------------------------------+-------+----------------+--------------------------------------------------------------------------------------------------------------+----------------+
| c:\users\administrator\desktop\tools\bginfo.exe | Grant | SYSTEM         | Specific Rights All,Delete,Read Control,Write DAC,Write Owner,Std Rights Required,Synchronize,Std Rights All | Inherited Ace  |
| c:\users\administrator\desktop\tools\bginfo.exe | Grant | Administrators | Specific Rights All,Delete,Read Control,Write DAC,Write Owner,Std Rights Required,Synchronize,Std Rights All | Inherited Ace  |
| c:\users\administrator\desktop\tools\bginfo.exe | Grant | Administrator  | Specific Rights All,Delete,Read Control,Write DAC,Write Owner,Std Rights Required,Synchronize,Std Rights All | Inherited Ace  |
+-------------------------------------------------+-------+----------------+--------------------------------------------------------------------------------------------------------------+----------------+
PS E:\O\osquery\b> .\osquery\RelWithDebInfo\osqueryi.exe "select * from ntfs_acl_permissions where path = 'c:\Windows\system32\notepad.exe'"
+---------------------------------+-------+-------------------------------------+--------------------------------------------------------------------------------------------------------------+----------------+
| path                            | type  | principal                           | access                                                                                                       | inherited_from |
+---------------------------------+-------+-------------------------------------+--------------------------------------------------------------------------------------------------------------+----------------+
| c:\Windows\system32\notepad.exe | Grant | TrustedInstaller                    | Specific Rights All,Delete,Read Control,Write DAC,Write Owner,Std Rights Required,Synchronize,Std Rights All | No Inheritance |
| c:\Windows\system32\notepad.exe | Grant | Administrators                      | Specific Rights All,Read Control,Std Rights Required,Synchronize,Std Rights All                              | No Inheritance |
| c:\Windows\system32\notepad.exe | Grant | SYSTEM                              | Specific Rights All,Read Control,Std Rights Required,Synchronize,Std Rights All                              | No Inheritance |
| c:\Windows\system32\notepad.exe | Grant | Users                               | Specific Rights All,Read Control,Std Rights Required,Synchronize,Std Rights All                              | No Inheritance |
| c:\Windows\system32\notepad.exe | Grant | ALL APPLICATION PACKAGES            | Specific Rights All,Read Control,Std Rights Required,Synchronize,Std Rights All                              | No Inheritance |
| c:\Windows\system32\notepad.exe | Grant | ALL RESTRICTED APPLICATION PACKAGES | Specific Rights All,Read Control,Std Rights Required,Synchronize,Std Rights All                              | No Inheritance |
+---------------------------------+-------+-------------------------------------+--------------------------------------------------------------------------------------------------------------+----------------+
```

### CODE REVIEW

File: osquery/tables/system/windows/ntfs_acl_permissions.cpp [hunk 1]
```diff
@@ -79,53 +117,18 @@ std::string accessPermsToStr(const unsigned long pmask) {
...
-    } else {
-      return wstringToString(name);
-    }
-  }
-  case TRUSTEE_IS_OBJECTS_AND_NAME:
-    // ptstrName member is a pointer to an OBJECTS_AND_NAME struct
-    return wstringToString(
-        reinterpret_cast<OBJECTS_AND_NAME_W*>(trustee.ptstrName)->ptstrName);
-  default:
+  auto r = LookupAccountSidA(
```
Comment (theopolis) [inline osquery/tables/system/windows/ntfs_acl_permissions.cpp:125]: Is there a particular reason we are not using `LookupAccountSidA`?

Heads up @farfella, IIUC we should stick to using wide-string API versions.

File: osquery/tables/system/windows/ntfs_acl_permissions.cpp [hunk 2]
```diff
@@ -148,50 +152,32 @@ QueryData genNtfsAclPerms(QueryContext& context) {
...
-                            ? accessModeIter->second
-                            : "Unknown";
-      auto interitedFromiter = kInheritanceToStr.find(aceItem->grfInheritance);
-      auto inheritedFrom = (interitedFromiter != kInheritanceToStr.end())
-                               ? interitedFromiter->second
-                               : "Unknown";
-      auto trusteeName = trusteeToStr(aceItem->Trustee);
+      // Get ACE
+      if (GetAce(dacl, cAce, (LPVOID*)&pAce) == FALSE) {
+        VLOG(1) << "GetAce failed. Error = " << GetLastError();
```
Comment (theopolis) [inline osquery/tables/system/windows/ntfs_acl_permissions.cpp:165]: nitpick, there's inconsistencies in how the verbose logging is happening, see above on line 155. I'd recommend stick to the precedent for now and use `"GetAce Error" << GetLastError()`.

**Review by theopolis — CHANGES_REQUESTED** [review]

File: osquery/tables/system/windows/ntfs_acl_permissions.cpp (hunk 1 above)
Comment (farfella) [reply osquery/tables/system/windows/ntfs_acl_permissions.cpp:125]: Yes. We should request changes if any -A calls are used.

File: osquery/tables/system/windows/ntfs_acl_permissions.cpp [hunk 3]
```diff
@@ -148,50 +152,32 @@ QueryData genNtfsAclPerms(QueryContext& context) {
...
       continue;
     }
 
-    // Loop through list of entries
-    auto aceItem = aceList;
-    for (unsigned long aceIndex = 0; aceIndex < aceCount;
-         aceItem++, aceIndex++) {
+    ACCESS_ALLOWED_ACE* pAce = NULL;
+    // Loop through the ACEs and display the information.
+    for (int cAce = 0; cAce < dacl->AceCount; cAce++) {
```
Comment (farfella) [inline osquery/tables/system/windows/ntfs_acl_permissions.cpp:161]: Should be WORD vs int since dacl->AceCount's type is WORD.

**Review by theopolis — APPROVED** [review]

**Pull request merged**